oak_runtime/src/node/wasm/mod.rs

```diff
@@ -634,6 +667,86 @@ impl WasmInterface {
             LabelReadStatus::NeedsCapacity(_) => Err(OakStatus::ErrBufferTooSmall),
         }
     }
+
+    /// Helper function to fetch a sequence of bytes from linear memory.
+    fn fetch_bytes(
+        &self,
+        bytes_ptr: AbiPointer,
+        bytes_length: AbiPointerOffset,
+    ) -> Result<Vec<u8>, OakStatus> {
+        self.get_memory()
+            .get(bytes_ptr, bytes_length as usize)
+            .map_err(|err| {
+                error!(
+                    "{}: fetch_bytes(): Unable to read name from guest memory: {:?}",
+                    self.pretty_name, err
+                );
+                OakStatus::ErrInvalidArgs
+            })
+    }
+
+    /// Helper function to fetch a [`String`] from linear memory.
+    fn fetch_string(
+        &self,
+        string_ptr: AbiPointer,
+        string_length: AbiPointerOffset,
+    ) -> Result<String, OakStatus> {
+        let bytes = self.fetch_bytes(string_ptr, string_length)?;
+
+        String::from_utf8(bytes).map_err(|err| {
+            error!(
+                "{}: fetch_string(): Unable to decode name as a string: {:?}",
+                self.pretty_name, err
+            );
+            OakStatus::ErrInvalidArgs
+        })
+    }
+
+    /// Helper function to fetch a proto from linear memory.
+    fn fetch_proto<T: prost::Message + Default>(
+        &self,
+        data_ptr: AbiPointer,
+        data_length: AbiPointerOffset,
+    ) -> Result<T, OakStatus> {
+        let bytes = self.fetch_bytes(data_ptr, data_length)?;
+
+        T::decode(bytes.as_slice()).map_err(|err| {
+            warn!(
+                "{}: fetch_data(): Could not parse node configuration: {:?}",
+                self.pretty_name, err
+            );
+            OakStatus::ErrInvalidArgs
+        })
+    }
+
+    /// Helper function to fetch bytes and handles from linear memory as a [`NodeMessage`].
+    fn fetch_message(
+        &self,
+        source: AbiPointer,
+        source_length: AbiPointerOffset,
+        handles: AbiPointer,
+        handles_count: AbiPointerOffset,
+    ) -> Result<NodeMessage, OakStatus> {
+        let bytes = self.fetch_bytes(source, source_length)?;
+
+        let raw_handles = self
+            .get_memory()
+            .get(handles, handles_count as usize * 8)
+            .map_err(|err| {
+                error!(
+                    "{}: fetch_message(): Unable to read handles from guest memory: {:?}",
+                    self.pretty_name, err
+                );
+                OakStatus::ErrInvalidArgs
+            })?;
+
+        let handles: Vec<u64> = raw_handles
+            .chunks(8)
+            .map(|bytes| LittleEndian::read_u64(bytes))
+            .collect();
```

Comment:
Given how neatly you've encapsulated things, maybe put this chunk into a `fetch_handles()` helper too?

Reply:
Done.